e2e/nomostest/portforwarder/port_forwarder.go

```diff
@@ -0,0 +1,230 @@
+	if async {
+		pf.mux.Lock()
+		defer pf.mux.Unlock()
+	}
+	pf.logger.Infof("updating port-forward %s:%d -> %s:%d", pf.pod, pf.localPort, pod, port)
+	if pf.setPortCallback != nil {
+		pf.setPortCallback(port, pod)
+	}
+	pf.localPort = port
+	pf.pod = pod
+}
+
+// portForwardToPod establishes a port forwarding to the provided pod name. This
+// should normally only be used as a helper function for start
+func (pf *PortForwarder) portForwardToPod(pod string, async bool) (*exec.Cmd, error) {
+	pf.logger.Infof("starting port-forward process for %s/%s %s", pf.ns, pod, pf.port)
+	cmd := exec.CommandContext(pf.ctx, "kubectl", "--kubeconfig", pf.kubeConfigPath, "port-forward",
+		"-n", pf.ns, pod, pf.port)
+
+	stdout := &strings.Builder{}
+	stderr := &strings.Builder{}
+	cmd.Stdout = stdout
+	cmd.Stderr = stderr
+
+	err := cmd.Start()
+	if err != nil {
+		return cmd, err
+	}
+	if stderr.Len() != 0 {
+		return cmd, fmt.Errorf(stderr.String())
+	}
+
+	localPort := 0
+	// In CI, 1% of the time this takes longer than 20 seconds, so 30 seconds seems
+	// like a reasonable amount of time to wait.
+	took, err := retry.Retry(30*time.Second, func() error {
+		select {
+		case <-pf.ctx.Done():
+			return pf.ctx.Err()
+		default:
+		}
+		s := stdout.String()
+		if !strings.Contains(s, "\n") {
+			return fmt.Errorf("nothing written to stdout for kubectl port-forward, stdout=%s", s)
+		}
+
+		line := strings.Split(s, "\n")[0]
+
+		// Sample output:
+		// Forwarding from 127.0.0.1:44043
+		_, err = fmt.Sscanf(line, "Forwarding from 127.0.0.1:%d", &localPort)
+		if err != nil {
+			return fmt.Errorf("unable to parse port-forward output: %q", s)
+		}
+		return nil
+	})
+	if err != nil {
+		return cmd, err
+	}
+	pf.logger.Infof("took %v to wait for port-forward to pod %s/%s (localhost:%d)", took, pf.ns, pod, localPort)
+	pf.setPort(localPort, pod, async)
+	return cmd, nil
+}
+
+// start creates a port forward to a deployment's pod and keeps it open,
+// even if the pod is recreated.
+// Once ctx is cancelled, the port forward will be closed and will no longer be recreated
+func (pf *PortForwarder) start() error {
+
+	pod, err := pf.kubeClient.GetDeploymentPod(pf.deployment, pf.ns, pf.retryTimeout)
+	if err != nil {
+		return err
+	}
+	cmd, portForwardErr := pf.portForwardToPod(pod.Name, false)
+	if portForwardErr != nil {
+		return portForwardErr
+	}
+	// Restart the port forward if the process exits prematurely (e.g. due to pod eviction)
+	go func() {
+		for {
+			if err := pf.watcher.WatchForNotFound(kinds.Pod(), pod.Name, pf.ns, testwatcher.WatchTimeout(1*time.Hour), testwatcher.WatchContext(pf.ctx)); err != nil {
+				if *e2e.Debug {
+					pf.logger.Info(err)
+				}
+			}
+			select {
+			case <-pf.ctx.Done():
+				return
+			default:
+			}
+			pf.logger.Info("stopping port-forward process")
+			if err := cmd.Process.Kill(); err != nil && errors.Is(err, os.ErrProcessDone) {
+				pf.logger.Info("error: killing port-forward process: %v", err)
```

Comment:
So i guess with the logger we can't kill the test, but it's still weird to have it log "errors". Should they be "warning"? Are we waiting for the next command that uses the port to fail to fail the tests? If so, can you leave a comment explaining that this doesn't need to be terminal because the next attempt to use the LocalPort will error?

Reply:
I originally had this calling `Fatal`, but refactored this when switching from `NT` to `Logger`. I decided to leave these as logs since we don't necessarily need to fail the test if the test is done using the port forward. But you are correct that the test would fail if it tried to use the port forward again after this failed. I can leave a comment

Reply:
left a comment